test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        # The choice of 32 maxconnections results in a maximum of 21 inbound connections
+        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
+        self.extra_args = [['-maxconnections=32']]
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def run_test(self):
+        self.protected_peers = set() # peers that we expect to be protected from eviction
+        current_peer = 0
+        node = self.nodes[0]
+        node.generatetoaddress(101, node.get_deterministic_priv_key().address)
+
+        # The first 4 peers send us a block, protecting them from eviction.
+        for _ in range(4):
+            block_peer = node.add_p2p_connection(SlowP2PDataStore())
+            current_peer += 1
+            block_peer.sync_with_ping(timeout=5)
+            best_block = node.getbestblockhash()
+            tip = int(best_block, 16)
+            best_block_time = node.getblock(best_block)['time']
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            block_peer.send_blocks_and_test([block], node, success=True)
+            self.protected_peers.add(current_peer)
+
+        # The next 5 nodes are slow-pinging peers, making them eviction candidates.
+        for _ in range(5):
+            node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+
+
+        #The next 4 peers send us a tx, protecting them from eviction
+        for i in range(4):
+            txpeer = node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+            txpeer.sync_with_ping(timeout=5)
+
+            prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
+            rawtx = node.createrawtransaction(
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
+            )
+            sigtx = node.signrawtransactionwithkey(
+                hexstring=rawtx,
+                privkeys=[node.get_deterministic_priv_key().key],
+                prevtxs=[{
+                    'txid': prevtx['txid'],
+                    'vout': 0,
+                    'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
+                }],
+            )['hex']
+            txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            self.protected_peers.add(current_peer)
+
+        # The next 8 peers have faster pings, which will usually protect them from eviction.
+        for _ in range(8):
+            fastpeer = node.add_p2p_connection(P2PInterface())
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
```

Comment:
This is probably not material, but the reason is that I wanted to trigger the eviction mechanism immediately after the `getpeerinfo` call, so that the node does not get time to exchange additional pings with its peers while the test does other things.

Reply:
I wondered the same thing. It's confusing that you do some of the processing before adding the new peer (`pings = [[...`) and some after  (`for i in range(...`). I think you should do all the processing before adding the new connection, since it'll take very little time, and the chance of there being another ping in between is ~zero.

If you do keep this ordering, I think it's worth a comment to say why you're doing it.

Reply:
I changed the ordering since the probability of furthers pings is extremely low.